### ISSUE TITLE
workers re-firing for webhook events that occurred 10+ days ago

### ISSUE THREAD

**Issue by Strandedpirate** [Azure/brigade#462]
We have a project setup with brigade and the bitbucket plugin. The last webhook/git check-in event occurred 10+ days ago for that project. For some reason brigade is firing off workers again for events/webhooks yesterday and again just now.

Any theories on why this would occur?
Any details I can provide to troubleshoot?

I've confirmed that bitbucket has NOT re-fired any old webhooks.

```
λ kubectl -n brigade get po -a -l component=build --sort-by=.status.startTime          
NAME                                        READY     STATUS       RESTARTS   AGE      
brigade-worker-01cbnkkskkwgm1gshd88pcgjdh   0/1       Init:Error   0          11d      
brigade-worker-01cc2nqw30d6egksqfkm3f5mf8   0/1       Completed    0          11d      
brigade-worker-01cc6vv27jknkmy3kq2t8x1275   0/1       Completed    0          11d      
brigade-worker-01cc6rnf7z7bcjjfxj1tcvfftz   0/1       Completed    0          11d      
brigade-worker-01ccjdxnrm2fz1mbtsyfcwty89   0/1       Completed    0          11d      
brigade-worker-01cc6xpz3f6zs9cv3b8jg8ew18   0/1       Completed    0          11d      
brigade-worker-01cc2tzsjvn11cvcq595ytx32x   0/1       Completed    0          11d      
brigade-worker-01cc81vnspzfsrbcaxp7ztpsdd   0/1       Completed    0          11d      
brigade-worker-01ccbmgvwfzhrpm83nmphqa1rj   0/1       Completed    0          11d      
brigade-worker-01ccja55681s2nww3apr7ahq56   0/1       Completed    0          11d      
brigade-worker-01cc2t5tpghy3vgc2cwmmy4x31   0/1       Completed    0          11d      
brigade-worker-01cc2jpdsk6fbnw35sk6mrqb99   0/1       Completed    0          11d      
brigade-worker-01cc6xk3rdexddqqxk31xp9mwr   0/1       Completed    0          11d      
brigade-worker-01cc82p2gr2b6whb5qc5v756pr   0/1       Completed    0          11d      
brigade-worker-01cc2fjn53dfe78cmxf3ear6kw   0/1       Completed    0          11d      
brigade-worker-01ccj9dkb21c43tz0mpbhp0bcm   0/1       Completed    0          11d      
brigade-worker-01cc6x7s30e3h3b1sd2525t1v0   0/1       Completed    0          11d      
brigade-worker-01cc6xcw78vwhhgt323wahckyt   0/1       Completed    0          11d      
brigade-worker-01ccxp1vn3b2zs9v85zejh5qv9   0/1       Completed    0          10d      
brigade-worker-01cc82cv4jyznb3atayyqttvnn   0/1       Completed    0          1d       
brigade-worker-01ccj9yzr70e9501ycc1bs8552   0/1       Completed    0          1d       
brigade-worker-01ccepv1eknwcx0eqb37t04acm   0/1       Init:Error   0          1d       
brigade-worker-01cbqjvkn2w9hxrstpc1zn1ejz   0/1       Init:Error   0          1d       
brigade-worker-01ccjdfxb7yt4h5s9ksxv0edk0   0/1       Completed    0          1d       
brigade-worker-01cc6w6f63yv4wy15revrp7j5d   0/1       Completed    0          1h       
brigade-worker-01cc80cbv36ge7t6r1zwhrf0qn   0/1       Completed    0          1h       
brigade-worker-01cc80t42fea3ftbspv5xqyfpd   0/1       Completed    0          1h       
brigade-worker-01cc813mw3aazd69kv063mjyts   0/1       Completed    0          1h       
brigade-worker-01cc6pq76y9f9gab1d7v4yvjb5   0/1       Completed    0          1h       
brigade-worker-01cc821wf7rseqvyzespzpjt6n   0/1       Completed    0          1h       
brigade-worker-01cc2qbhfpw36djpj50qybe5fa   0/1       Completed    0          1h       
brigade-worker-01cc80mp0k4pkx9drca7nmndd7   0/1       Completed    0          1h       
brigade-worker-01ccjdrf6jc019ex3t473bshk6   0/1       Completed    0          1h       
brigade-worker-01ccm03ds5nn3m6c8vczgm3bwq   0/1       Completed    0          1h       
brigade-worker-01cc6wc8p8a6vw2xbjm4927kbx   0/1       Completed    0          1h       
brigade-worker-01cbnk3mspgej3fnkd39twc8tp   0/1       Init:Error   0          1h       
brigade-worker-01ccj92kxan2z4g08r47jk3eym   0/1       Completed    0          1h       
brigade-worker-01cc2sey3ja6cghpzrfp7fe9jh   0/1       Completed    0          1h       
brigade-worker-01cc6xv00f05t266wzk4jdqm9q   0/1       Completed    0          1h       
brigade-worker-01cc1rw508v2s0m4710z3jbxna   0/1       Init:Error   0          1h       
brigade-worker-01cc828maz379y1m01yps63tzt   0/1       Completed    0          1h       
brigade-worker-01cc8108nz4yx1xcfzdr1njpqg   0/1       Completed    0          1h       
brigade-worker-01cch73y4v90sskhk60jf4r0k6   0/1       Completed    0          1h       
brigade-worker-01cbqvbyqkcsw92pb1yq3e4kp8   0/1       Init:Error   0          1h       
brigade-worker-01cc6wrttpmw5eb75nmejh4373   0/1       Completed    0          1h       
brigade-worker-01cc2gmv6scpm40sfbrqbwcvkb   0/1       Completed    0          1h       
brigade-worker-01cc2rfqthneq5xa6hhk2r42an   0/1       Completed    0          1h       
brigade-worker-01cc2v1k6xtq0grvjxdcg3rfxw   0/1       Completed    0          1h       
brigade-worker-01cc6wy9d07vfbyvx2hz7mdm25   0/1       Completed    0          1h       
brigade-worker-01cbnkdpxpc1pmys1z7wrpb6s0   0/1       Init:Error   0          1h       
brigade-worker-01cc1vddvynjrzhyrhvsh7f1qj   0/1       Init:Error   0          1h       
brigade-worker-01cc6x55r3am8402ec7cgpmggw   0/1       Completed    0          1h       
brigade-worker-01cc7zwm4jjxh68dr2v5ay2fnk   0/1       Completed    0          1h       
brigade-worker-01ccjcvrqteh1kzy7q0kbd0tcv   0/1       Completed    0          1h       
```

```
λ brig build -n brigade get 01cbnkdpxpc1pmys1z7wrpb6s0                           
id: 01cbnkdpxpc1pmys1z7wrpb6s0                                                   
projectid: brigade-f747adf4abe234fe7d7a399df33ee2cffe7f66b7643438bbd32d72        
type: push                                                                       
provider: bitbucket                                                              
revision:                                                                        
  commit: 11f5444171d1b8f8a04a0b5d5e07722f382bee0a                               
  ref: ""                                                                        
payload: []                                                                      
script: []                                                                       
worker:                                                                          
  id: brigade-worker-01cbnkdpxpc1pmys1z7wrpb6s0                                  
  buildid: 01cbnkdpxpc1pmys1z7wrpb6s0                                            
  projectid: brigade-f747adf4abe234fe7d7a399df33ee2cffe7f66b7643438bbd32d72      
  starttime: 2018-05-17T10:48:52-05:00                                           
  endtime: 0001-01-01T00:00:00Z                                                  
  exitcode: 0                                                                    
  status: Failed                                                                 
loglevel: ""    
```                 
The last 15 commits for the project:
```
λ git log --pretty=format:"%H %ad" -n 15 --date=rfc                                       
e5e7822a770419e2a46e8f73edd92144b18702ef Mon, 7 May 2018 11:13:10 -0500           
688b21798535d2d83a587e6d530965b57a6ecb93 Thu, 3 May 2018 16:56:23 -0500           
de1b4ff9dbca9415a4de7a58e4ae4b874f75a13f Thu, 3 May 2018 02:19:29 -0500           
99b7f0732777c554ed9f0427b15284685ca0e878 Thu, 3 May 2018 02:16:33 -0500           
f22bd9a5fc83c7aa165c1f1a98b2d11f841c3192 Thu, 3 May 2018 02:11:56 -0500           
dabf1131108f1412533077d06fc14ad9361a3969 Thu, 3 May 2018 02:00:53 -0500           
72442d6eb951cfd83d11e35d7390e85b435dd43d Thu, 3 May 2018 01:13:39 -0500           
816635e64b6590e71c915787d6ae1330dfa7fdda Thu, 3 May 2018 01:10:17 -0500           
60b82a97e235b16ec27e12fb51737382be957f0f Thu, 3 May 2018 01:00:47 -0500           
8b5012b97a3cda3204574fe1220eb27e3676b243 Thu, 3 May 2018 00:54:47 -0500           
928346b46f4792592b5705e10b1e11157cd4d1a0 Thu, 3 May 2018 00:30:10 -0500           
0ecfea0e0bbc68c78359de5ba81873c1497447e3 Wed, 2 May 2018 15:01:13 -0500           
b4f418a6f171f875959a9105ca20bbb448ea53fa Tue, 1 May 2018 15:38:20 -0500           
3737d3a94b85df86540ad7ad87d487e2af5a0cd3 Mon, 30 Apr 2018 11:00:06 -0500                               
```

**Comment by Strandedpirate** [comment]
Yeah, looks like the controller, api, and gateway pods were started yesterday. I can't find an old replicaset or previous pod for them which is abnormal. There is also no previous state nor exit code marked in the pod's yaml below. It's as if these pods never existed until the other day; but I know they did..

Would updating a brigade-project cause this behavoir? We haven't touched brigade otherwise.
`helm upgrade -i --namespace brigade brigade-designer brigade/brigade-project -f brigade-project.yaml`

Is there nothing in the code-base that marks events, stored as secrets, as being processed so that it's impossible to process them a second time?

Since all pods are mortal we must assume every pod in the system is going to die any number of times and be restarted and the system should not attempt to re-process events from the beginning of time.

```
λ kubectl -n brigade get rs,po -a
NAME                                            DESIRED   CURRENT   READY     AGE
rs/bb-gw-brigade-bitbucket-gateway-69cd57fbfb   0         0         0         26d
rs/bb-gw-brigade-bitbucket-gateway-7dd68647ff   1         1         1         21d
rs/brigade-brigade-api-7f79c58ddd               1         1         1         21d
rs/brigade-brigade-ctrl-77bf5f7bcb              1         1         1         21d
rs/brigade-brigade-github-gw-568845f6b5         1         1         1         21d

NAME                                                  READY     STATUS       RESTARTS   AGE
po/bb-gw-brigade-bitbucket-gateway-7dd68647ff-d45xf   1/1       Running      0          1d
po/brigade-brigade-api-7f79c58ddd-2nkc5               1/1       Running      0          1d
po/brigade-brigade-ctrl-77bf5f7bcb-7nkdr              1/1       Running      0          1d
po/brigade-brigade-github-gw-568845f6b5-n67l6         1/1       Running      2          12d
...
```

controller pod:
```
λ kubectl -n brigade get po brigade-brigade-ctrl-77bf5f7bcb-7nkdr -o yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubernetes.io/created-by: |
      {"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"ReplicaSet","namespace":"brigade","name":"brigade-brigade-ctrl-77bf5f7bcb","uid":"95b00306-49ca-11e8-bf32-000d3a755dcb","apiVersion":"extensions","resourceVersion":"34053149"}}
  creationTimestamp: 2018-05-17T15:48:32Z
  generateName: brigade-brigade-ctrl-77bf5f7bcb-
  labels:
    app: brigade-brigade-ctrl
    pod-template-hash: "3369193676"
    role: controller
  name: brigade-brigade-ctrl-77bf5f7bcb-7nkdr
  namespace: brigade
  ownerReferences:
  - apiVersion: extensions/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: brigade-brigade-ctrl-77bf5f7bcb
    uid: 95b00306-49ca-11e8-bf32-000d3a755dcb
  resourceVersion: "34053841"
  selfLink: /api/v1/namespaces/brigade/pods/brigade-brigade-ctrl-77bf5f7bcb-7nkdr
  uid: c06788ee-59e9-11e8-9b18-000d3a755dcb
spec:
  containers:
  - env:
    - name: BRIGADE_NAMESPACE
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.namespace
    - name: BRIGADE_WORKER_IMAGE
      value: deis/brigade-worker:v0.13.1
    - name: BRIGADE_WORKER_PULL_POLICY
      value: IfNotPresent
    - name: BRIGADE_WORKER_SERVICE_ACCOUNT
      value: brigade-worker
    image: deis/brigade-controller:v0.13.1
    imagePullPolicy: IfNotPresent
    name: brigade
    ports:
    - containerPort: 7744
      protocol: TCP
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: brigade-brigade-ctrl-token-ldlxr
      readOnly: true
  dnsPolicy: ClusterFirst
  nodeName: k8s-agent-732ff0bd-3
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: brigade-brigade-ctrl
  serviceAccountName: brigade-brigade-ctrl
  terminationGracePeriodSeconds: 30
  volumes:
  - name: brigade-brigade-ctrl-token-ldlxr
    secret:
      defaultMode: 420
      secretName: brigade-brigade-ctrl-token-ldlxr
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: 2018-05-17T15:48:32Z
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: 2018-05-17T15:48:33Z
    status: "True"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: 2018-05-17T15:48:32Z
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: docker://2d72d3fd2b5e2020730f03bea83a2967d9033511a7ad93644659f22eb4f36d08
    image: deis/brigade-controller:v0.13.1
    imageID: docker-pullable://deis/brigade-controller@sha256:94b39d69b12e8c7537393554d4f1e4f1bda59258d7008d2d1ed4e3ae45b847bd
    lastState: {}
    name: brigade
    ready: true
    restartCount: 0
    state:
      running:
        startedAt: 2018-05-17T15:48:33Z
  hostIP: 10.240.0.6
  phase: Running
  podIP: 10.244.4.250
  qosClass: BestEffort
  startTime: 2018-05-17T15:48:32Z
```

**Comment by radu-matei** [comment]
I think this was fixed in #482.
@adamreese, @technosophos - could you please confirm?
Thanks!

**Comment by technosophos** [comment]
This was fixed in #482, though older builds might run once more on upgrade (because they don't have the label). So as of 0.15, this issue is considered resolved.
